proposals/type-system/schema2owl.js

```diff
@@ -0,0 +1,230 @@
+'use strict'
+
+const ctx = {
+    'owl': 'http://www.w3.org/2002/07/owl#',
+    'xsd': 'http://www.w3.org/2001/XMLSchema#',
+    'boolean': 'xsd:boolean',
+    'integer': 'xsd:integer',
+    'number': 'xsd:double', // TODO fix this
+    'string': 'xsd:string',
+    'allOf': {
+        '@id': 'owl:intersectionOf',
+        '@type': '@id',
+        '@container': '@list'
+    },
+    // TODO JSON Schema's oneOf (vs. anyOf?)
+    'anyOf': {
+        '@id': 'owl:unionOf',
+        '@type': '@id',
+        '@container': '@list'
+    },
+    'enum': {
+        '@id': 'owl:oneOf',
+        '@type': '@id',
+        '@container': '@list'
+    },
+    'withRestrtictions': {
+        '@id': 'owl:withRestrtictions',
+        '@type': '@id',
+        '@container': '@list'
+    },
+    'equivalentClass': {
+        '@id': 'owl:equivalentClass',
+        '@type': '@id'
+    },
+    'onDatatype': {
+        '@id': 'owl:onDatatype',
+        '@type': '@id'
+    },
+    'onProperty': {
+        '@id': 'owl:onProperty',
+        '@type': '@id'
+    },
+    'allValuesFrom': {
+        '@id': 'owl:allValuesFrom',
+        '@type': '@id'
+    },
+    'someValuesFrom': {
+        '@id': 'owl:someValuesFrom',
+        '@type': '@id'
+    },
+    'minimum': {
+        '@id': 'xsd:minInclusive'
+    },
+    'maximum': {
+        '@id': 'xsd:maxInclusive'
+    },
+    'exclusiveMinimum': {
+        '@id': 'xsd:minExclusive'
+    },
+    'exclusiveMaximum': {
+        '@id': 'xsd:maxExclusive'
+    },
+    'minLength': {
+        '@id': 'xsd:minLength'
+    },
+    'maxLength': {
+        '@id': 'xsd:maxLength'
+    },
+    'pattern': {
+        '@id': 'xsd:pattern'
+    }
+}
+
+function canonicalize(json, root, stack, base) {
+    if (json['$ref']) {
+        let refStack = json['$ref'].split('/');
+        let r = refStack.shift(); // remove root
+        if (r === '#') { // local definition
+            let ref = refStack.reduce((node, idx) => node[idx], root);
+            return canonicalize(ref, root, refStack, base);
+        } else {
+            // FIXME specific to OCF
+            let refBase = base.substring(0, base.lastIndexOf('/') + 1);
+            return { 'id': refBase + json['$ref'] };
+        }
+    } else {
+        if (!json.id) {
```

Comment:
My apologies if I'm misinterpreting this.  I'm assuming this is intended to support JSON Schema draft-04?  In draft-06, the `id` keyword became `$id` (to match `$schema` and `$ref`, and because people constatly confused it with instance properties named `id`)

Reply:
Ah. Thanks for the information. As you might have seen in the code, I tested only against the JSON Schema definitions from https://oneiota.org/, which are based on draft-04.

It is a good thing, indeed to have changed that keyword!